Kerberos.NET/Client/Krb5TicketCache.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Kerberos.NET.Client
+{
+    [DebuggerDisplay("{cache}")]
+    public class Krb5TicketCache : TicketCacheBase
+    {
+        private readonly string filePath;
+        private readonly Krb5CredentialCache cache;
+
+        private readonly object fileSync = new object();
+
+        public Krb5TicketCache(string filePath, ILoggerFactory logger = null)
+            : this(logger)
+        {
+            this.filePath = filePath;
+            this.ReadCache();
+        }
+
+        public Krb5TicketCache(byte[] cache, ILoggerFactory logger = null)
+            : this(logger)
+        {
+            this.ReadCache(cache);
+        }
+
+        protected Krb5TicketCache(ILoggerFactory logger)
+            : base(logger)
+        {
+            this.cache = new Krb5CredentialCache();
+            this.cache.Header[Krb5CredentialCacheTag.KdcClientOffset] = new byte[8];
+        }
+
+        private void ReadCache(byte[] cache)
+        {
+            if (cache == null || cache.Length <= 0)
+            {
+                return;
+            }
+
+            this.cache.Read(cache);
+        }
+
+        private void ReadCache()
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                return;
+            }
+
+            using (var file = this.OpenFile())
+            {
+                var cache = new byte[file.Length];
+
+                int offset = 0, read = 0;
+
+                do
+                {
+                    read = file.Read(cache, offset, cache.Length - offset);
+
+                    offset += read;
+                }
+                while (read > 0);
+
+                this.ReadCache(cache);
+            }
```

Comment:
```suggestion
            byte[] cache = File.ReadAllBytes(filePath);
            this.ReadCache(cache);
```
Is there a reason for the explicit read-loop?

Reply:
The explicit FileMode.OpenOrCreate, FileAccess.ReadWrite, and FileShare.None in this.OpenFile(); AFAIK ReadAllBytes/WriteAllBytes don't use these modes so there's a risk of two reads or writes colliding from another process. This is more likely to happen on *nix systems running MIT or Heimdal side by side and sharing the same cache file. If ReadAll/WriteAll do this, I'm happy to change this though.

Reply:
That's a good point, and you're right that [File.ReadAllBytes](https://github.com/dotnet/runtime/blob/aa5fdab9654d74bc6274c0b5d820272c8e859621/src/libraries/System.IO.FileSystem/src/System/IO/File.cs#L331) uses different flags.

Maybe it's worth to refactor this read/write out of this class?
It's not the (primary) concern of `Krb5TicketCache` to do this.

Reply:
It actually is the primary concern. The parsing of the format is done in internal Krb5CredentialCache, so public Krb5TicketCache can be the glue that binds it to a cache file and lets it be usable to the KerberosClient.

Reply:
Yeah, that's true. But for me reading and writing from/to the file is a different concern than the parsing, so I'd split this out in a `FileHelper`, etc. Though that's a quibble 😉 
I'm fine when you leave it as is.